packages/jest-transform/src/ScriptTransformer.ts

```diff
@@ -467,9 +475,28 @@ export default class ScriptTransformer {
     );
     const module: ModuleType = require(moduleName);
 
+    if (!callback) {
+      revertHook();
+
+      return module;
+    }
+
     try {
-      if (callback) {
-        await callback(module);
+      const cbResult = callback(module);
+
+      if (isPromise(cbResult)) {
+        return cbResult.then(
```

Comment:
The workaround for not being able to use finally here is pretty ugly IMO, can we use an async function with try-finally like before, but only for the asynchronous case? Can even be an async IIFE.
Also, the existing try-finally can be shorter to only cover the synchronous case.

Reply:
what about now?